### PR TITLE
Do not stop daemon from booting if io.EOF on loading image

### LINKED ISSUE #17618: [1.9.0-rc4] Can't survive a hard reboot
This has been a long standing issue with Docker but I haven't pushed on it because it didn't seem to happen very often.  With 1.9 it has gotten much worse and happens quite easily.

Docker writes a lot of little json files but it seems to do it in a way that is not very safe and has the ability to create zero bytes files quite easily.  On start Docker will attempt to read the 0 byte file and fail and then the daemon won't start.  Again in 1.8 this wasn't very often, in 1.9 it's is quite common that I see this.

Here is my scenario:

1. Start RancherOS VM from a clean disk image (using KVM)
2. Once booted run `docker run -itd busybox`
3. Kill VM
4. Next start Docker daemon will refused to start.  The error is below

```
time="2015-11-02T18:12:08.919123488Z" level=info msg="API listen on /var/run/docker.sock" 
time="2015-11-02T18:12:08.922337393Z" level=fatal msg="Error starting daemon: EOF" 
```

Looking at `/var/lib/docker` I have the following zero byte files

```
/var/lib/docker/graph/039b63dd2cbaa10d6015ea574392530571ed8d7b174090f032211285a71881d0/tar-data.json.gz
/var/lib/docker/graph/039b63dd2cbaa10d6015ea574392530571ed8d7b174090f032211285a71881d0/json
/var/lib/docker/graph/039b63dd2cbaa10d6015ea574392530571ed8d7b174090f032211285a71881d0/layersize
/var/lib/docker/graph/039b63dd2cbaa10d6015ea574392530571ed8d7b174090f032211285a71881d0/checksum
/var/lib/docker/graph/039b63dd2cbaa10d6015ea574392530571ed8d7b174090f032211285a71881d0/v1Compatibility
/var/lib/docker/graph/c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7/tar-data.json.gz
/var/lib/docker/graph/c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7/json
/var/lib/docker/graph/c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7/layersize
/var/lib/docker/graph/c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7/checksum
/var/lib/docker/graph/c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7/v1Compatibility
/var/lib/docker/containers/14c8e08438701d5143073531e81a273bc50b899469e5fe5558e45a2a4d8eae71/14c8e08438701d5143073531e81a273bc50b899469e5fe5558e45a2a4d8eae71-json.log
/var/lib/docker/containers/14c8e08438701d5143073531e81a273bc50b899469e5fe5558e45a2a4d8eae71/hostname
```

This exact scenario in 1.8 does not fail.  I don't know exactly what file is being read that is failing, but again this is very easy to reproduce and fails the Docker daemon from starting.

It would seem to me that docker needs to write to a temp file and then rename to dest to avoid these 0 byte file issues.

### PR DESCRIPTION
**Pull request by tiborvass**
Signed-off-by: Tibor Vass tibor@docker.com

Fixes #17618

Ping @calavera @runcom @ibuildthecloud

### CODE REVIEW

**Comment by calavera** [comment]
LGTM. We want to catch other errors that are not related to file corruption.

**Comment by tiborvass** [comment]
`TestInspectApiBridgeNetworkSettings120` is failing on windows, flakey?

**Comment by calavera** [comment]
I don't think it's related to this change, since the other builds are green.

File: graph/graph.go [hunk 1]
```diff
@@ -182,7 +182,10 @@ func (graph *Graph) restore() error {
 		if graph.driver.Exists(id) {
 			img, err := graph.loadImage(id)
 			if err != nil {
-				return fmt.Errorf("could not restore image %s: %v", id, err)
+				if err != io.EOF {
+					return fmt.Errorf("could not restore image %s: %v", id, err)
+				}
```
Comment (LK4D4) [inline graph/graph.go]: there should be log message at least

Comment (tiborvass) [inline graph/graph.go]: lol, too bad the CI is green

**Comment by LK4D4** [comment]
LGTM

**Comment by tiborvass** [comment]
CI was already green, merging #yolo self-five

**Comment by jfrazelle** [comment]
\o/

**Comment by calavera** [comment]
:pray:

**Comment by runcom** [comment]
I guess I'm late now :(

**Comment by thaJeztah** [comment]
Yeah, us Europeans miss all the fun booohoooo :smile:

**Pull request merged**